rust/experimental/lang/src/main.rs

```diff
@@ -14,142 +14,277 @@
 
 //! A language syntax coloring and indentation plugin for xi-editor.
 
-#[macro_use]
+extern crate xi_core_lib;
 extern crate xi_plugin_lib;
+extern crate xi_rope;
+extern crate xi_trace;
 
-use xi_plugin_lib::caching_plugin::{self, PluginCtx, SpansBuilder};
+use std::{collections::HashMap, env, path::Path};
 
-use std::env;
+use language::{plaintext::PlaintextParser, rust::RustParser};
+use parser::Parser;
+use statestack::{HolderNewState, State};
+use xi_core_lib::{plugins::rpc::ScopeSpan, ConfigTable, LanguageId, ViewId};
+use xi_plugin_lib::{mainloop, Cache, Plugin, StateCache, View};
+use xi_rope::RopeDelta;
+use xi_trace::{trace, trace_block, trace_payload};
 
-mod rust;
-mod colorize;
-mod statestack;
+mod language;
+mod parser;
 mod peg;
+mod statestack;
+
+const LINES_PER_RPC: usize = 50;
+
+type Scope = Vec<String>;
 
-use rust::RustColorize;
-use statestack::State;
-use colorize::{Style, StyleNewState, Colorize};
+struct LangPlugin {
+    view_states: HashMap<ViewId, ViewState>,
+}
 
-fn add_style_span(builder: &mut SpansBuilder, style: &Style, start: usize, end: usize) {
-    builder.add_style_span(start, end,
-        style.fg_color, style.font);
+impl LangPlugin {
+    fn new() -> LangPlugin {
+        LangPlugin { view_states: HashMap::new() }
+    }
 }
 
-struct PluginState {
-    colorize: RustColorize<StyleNewState<fn(&mut Style, &rust::StateEl)>>,
-    line_num: usize,
+impl Plugin for LangPlugin {
+    type Cache = StateCache<State>;
+
+    fn update(
+        &mut self,
+        view: &mut View<Self::Cache>,
+        _delta: Option<&RopeDelta>,
+        _edit_type: String,
+        _author: String,
+    ) {
+        let _guard = trace_block("ExperimentalLang::update", &["experimental-lang"]);
```

Comment:
Each trace we record increases the overall size and possibly sample count (we only keep 100k events per source, maybe?) so while tracing is definitely good it's also good to be judicious. Since this call should always return immediately there's not much to learn here; it's more important to trace things that have variable runtimes or that we're otherwise suspicious of, performance wise.

Reply:
Yeah, good point. Chatty logs (traces?) were helpful while actually getting this working, but we can cut down on most of them now. I'm going to remove all of the ones in the lifecycle methods, but I'll probably stick one in `compute_syntax`.